doc/source/serve/model_composition.md

```diff
@@ -212,65 +212,52 @@ $ python arithmetic_client.py
 9
 ```
 
-(deployment-graph-call-graph-testing)=
-### Testing the Call Graph with the Python API
-
-All `MethodNodes` and `FunctionNodes` have an `execute` method. You can use this method to test your graph in Python, without using HTTP requests. 
+(deployment-graph-drivers-http-adapters)=
+### Drivers and HTTP Adapters
 
-To test your graph,
+Ray Serve provides the `DAGDriver`, which routes HTTP requests through your call graph. As mentioned in [the call graph section](deployment-graph-call-graph), the `DAGDriver` takes in a `DeploymentNode` and it produces a `ClassNode` that you can run.
 
-1. Call `execute` on the `MethodNode` or `FunctionNode` that you would pass into the `DAGDriver`.
-2. Pass in the input to the graph as the argument. **This argument becomes the input represented by `InputNode`**. Make sure to refactor your call graph accordingly, since it takes in this input directly, instead of an HTTP request.
-3. `execute` returns a reference to the result, so the graph can execute asynchronously. Call `ray.get` on this reference to get the final result.
+The `DAGDriver` also has an optional keyword argument: `http_adapter`. [HTTP adapters](serve-http-adapters) are functions that get run on the HTTP request before it's passed into the graph. Ray Serve provides a handful of these adapters, so you can rely on them to conveniently handle the HTTP parsing while focusing your attention on the graph itself.
 
-As an example, we can rewrite the [arithmetic call graph example](deployment-graph-arithmetic-graph) from above to use `execute`:
+For instance, we can use the Ray Serve-provided `json_request` adapter to simplify our [arithmetic call graph](deployment-graph-arithmetic-graph) by eliminating the `unpack_request` function. You can replace lines 29 through 38 with this graph:
 
-```python
-with InputNode() as request_number:
-    add_2_output = add_2.add.bind(request_number)
-    subtract_1_output = subtract_one_fn.bind(add_2_output)
-    add_3_output = add_3.add.bind(subtract_1_output)
-
-ref = add_3_output.execute(5)
-result = ray.get(ref)
-print(result)
+(http-adapter-arithmetic-example)=
+```{literalinclude} doc_code/model_composition/arithmetic.py
+:start-after: __adapter_graph_start__
+:end-before: __adapter_graph_end__
+:language: python
 ```
 
-Then we can run the script directly:
-
-```
-$ python arithmetic.py
+Note that the `http_adapter`'s output type becomes what the `InputNode` represents. Without the `json_request` adapter, the `InputNode` represented an HTTP request. With the adapter, it now represents the number packaged inside the request's JSON body. You can work directly with that body's contents in the graph instead of first processing it.
 
```

Comment:
`InputNode` represents or `InputNode` expects? The InputNote is crucial so we need to make it clear: What it represents, what it does, what it expects with and without http_adapters. I wonder if that should be part of where we describe each type of inference graph node.

Reply:
The `InputNode` is a representation of the input at runtime. You can't manipulate the `InputNode` directly in the context manager. I added more details about how the `InputNode` works.

Reply:
yes, developers how to understand the semantics on `InputNode`, especially if it represents the input to their entire DAG.